types/index.d.ts

```diff
@@ -0,0 +1,157 @@
+declare module 'react-sweet-state' {
+  import { ComponentType, ReactNode, ReactElement } from 'react';
+
+  type SetState<TState> = (newState: Partial<TState>) => void;
+  type GetState<TState> = () => Readonly<TState>;
+  type StoreUnsubscribe = () => void;
+
+  type RenderPropComponent<TState, TActions> = (
+    state: TState,
+    actions: BoundActions<TActions>
+  ) => ReactNode;
+  type HookReturnValue<TState, TActions> = [TState, BoundActions<TActions>];
+
+  export type Store<TState, TActions> = {
+    key: string[];
+    initialState: TState;
+    actions: TActions;
+  };
+
+  export type StoreState<TState> = {
+    getState: GetState<TState>;
+    setState: SetState<TState>;
+    key: string[];
+    subscribe: (listener: () => void) => StoreUnsubscribe;
+    mutator: SetState<TState>;
+  };
+
+  export type Action<TState, TContainerProps = void, TActions = any> = (
+    state: {
+      setState: SetState<TState>;
+      getState: GetState<TState>;
+      actions: TActions;
+      dispatch: (actionThunk: Action<TState, TContainerProps, TActions>) => any;
+    },
+    containerProps: TContainerProps
+  ) => any;
+
+  type BoundActions<TActions> = TActions;
+
+  export interface StoreInstance<TState, TActions> {
+    store: StoreState<TState>;
+    actions: TActions;
+  }
+
+  export class Registry {
+    configure(options: { initialStates?: { [key: string]: Object } }): void;
+    stores: Map<string, StoreInstance<any, any>>;
+    initStore: <TState, TActions>(
+      store: Store<TState, TActions>,
+      key: string
+    ) => StoreInstance<TState, TActions>;
+    getStore: <TState, TActions>(
+      store: Store<TState, TActions>,
+      scopeId: string
+    ) => StoreInstance<TState, TActions>;
+    deleteStore: <TState, TActions>(
+      store: Store<TState, TActions>,
+      scopeId: string
+    ) => void;
+  }
+
+  export var defaultRegistry: Registry;
+
+  type MiddlewareResult = any;
+  export type Middleware = (
+    store: StoreState<any>
+  ) => (
+    next: (fn: any) => MiddlewareResult
+  ) => (fn: () => any) => MiddlewareResult;
+
+  export var defaults: {
+    devtools: boolean;
+    middlewares: any;
+    mutator: <TState>(
+      prevState: TState,
+      partialState: Partial<TState>
+    ) => TState;
+  };
+
+  type ContainerComponent<TProps> = ComponentType<
```

Comment:
I don't think this is the solution but for the moment it should work (might spark the correct idea). This type is also conflictive. The fact that we explicitly state `& TProps` implies that a possible `void` which requires props to be present on the component, with or without data. 

To avoid this we can type it as follows:
```
    type ContainerComponent<TProps> = ComponentType<
        | {
              scope?: string;
              isGlobal?: boolean;
          }
        | ({
              scope?: string;
              isGlobal?: boolean;
          } & TProps)
    >;
```

Reply:
I believe the solution is simpler: the default should be `{}` instead of `void`. 
```
export function createContainer<TState, TActions, TProps = {}>(...)
```